src/mlpack/core/cv/metrics/roc_auc_score_impl.hpp

```diff
@@ -0,0 +1,78 @@
+/**
+ * @file core/cv/metrics/roc_auc_score_impl.hpp
+ * @author Sri Madhan M
+ *
+ * Implementation of the area under Receiver Operating Characteristic curve
+ * (ROC-AUC) score.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_CV_METRICS_ROCAUCSCORE_IMPL_HPP
+#define MLPACK_CORE_CV_METRICS_ROCAUCSCORE_IMPL_HPP
+
+#include <mlpack/core/cv/metrics/accuracy.hpp>
+#include <iostream>
+
+namespace mlpack {
+namespace cv {
+
+template<size_t PC /* PositiveClass */>
+template<typename MLAlgorithm, typename DataType>
+double ROC_AUC<PC>::Evaluate(MLAlgorithm& model,
+                              const DataType& data,
+                              const arma::Row<size_t>& labels)
+{
+  util::CheckSameSizes(data, labels, "ROC_AUC::Evaluate()");
+
+  arma::mat probabilities;
+  model.Classify(data, probabilities);
+
+  // compute labels with "1" for positive class and "0" for the other
+  arma::Col<size_t> binaryLabels = arma::conv_to<arma::Col<size_t>>::from(
+                                        ((arma::umat) (labels == PC)).row(0));
+
+  // probability scores of positive class
+  arma::Col<double> scoresOfPC   = arma::conv_to<arma::Col<double>>::from(
+                                        probabilities.row(PC));
+
+  size_t numberOfTrueLabels  = arma::sum(binaryLabels);
+  size_t numberOfFalseLabels = binaryLabels.n_rows - numberOfTrueLabels;
+
+  // sort labels and probabilities, using probability scores
+  arma::ucolvec sortedScoreIndices = arma::stable_sort_index(
+                                        scoresOfPC, "descend");
+  arma::Col<size_t> sortedLabels = binaryLabels(sortedScoreIndices);
+  arma::Col<double> sortedScores = scoresOfPC(sortedScoreIndices);
+
+  // compute indices of unique probability scores
+  arma::ucolvec uniqueScoreIndices = arma::find(arma::diff(sortedScores));
+  uniqueScoreIndices.insert_rows(uniqueScoreIndices.n_rows, 1);
+  uniqueScoreIndices(uniqueScoreIndices.n_rows - 1) = scoresOfPC.n_rows - 1;
```

Comment:
This is a clever approach, but I'm not convinced it would be empirically faster than computing the ROC points by simply iterating over `sortedScores` and `sortedLabels`.  Also, I suspect that implementation might be easier to read.

Feel free to try it if you like on some very large example data, but my intuition is that `arma::find(arma::diff(sortedScores))` will not end up resulting in an efficient way of computing the "change points" of the probabilities.

Reply:
I compared the performance of `arma::find(arma::diff(sortedScores))` and its equivalent for loop version. Your thoughts are correct, for loop version is better. I will make the necessary changes.